sdks/python/apache_beam/utils/sharded_key.py

```diff
@@ -0,0 +1,65 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pytype: skip-file
+
+from __future__ import absolute_import
+
+
+class ShardedKey(object):
+  """A sharded key consisting of a user key and a shard id.
+
+  Attributes:
+    key: The user key.
+    shard_id: An opaque byte string that uniquely represents a shard of the key.
+  """
+  def __init__(
+      self,
+      key,
+      shard_id=b'',  # type: bytes
+  ):
+    # type: (...) -> None
+    assert shard_id is not None
+    self._key = key
+    self._shard_id = shard_id
+
+  @property
+  def key(self):
+    return self._key
+
+  @property
+  def shard_id(self):
```

Comment:
Do we intend this to be public? IMHO, it'd be better to not have it part of the public API (and we can always add it later if needed).

Reply:
Same for Java.

Reply:
At least the coder impl needs to access the shard id for encoding. In Java, the coder is defined inside of the ShardedKey class so we could make the shard id protected (not private for the AutoValue to work). But in Python the coder is defined separately.

Reply:
We can simply access _shard_id directly from the coder and get rid of this "public" accessor.